xds/internal/xdsclient/callback.go

```diff
@@ -80,37 +80,41 @@ func (c *clientImpl) NewListeners(updates map[string]ListenerUpdate, metadata Up
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
+	c.ldsVersion = metadata.Version
 	if metadata.ErrState != nil {
-		// On NACK, update overall version to the NACKed resp.
 		c.ldsVersion = metadata.ErrState.Version
-		for name := range updates {
-			if s, ok := c.ldsWatchers[name]; ok {
+	}
+	for name, update := range updates {
+		if s, ok := c.ldsWatchers[name]; ok {
+			if update.Err != nil {
```

Comment:
Would it be better to add a `PerResourceErrors map[string]error` field to `UpdateErrorMetadata` instead of adding the error directly to each `XxxUpdate` struct? We already have such a map ready in `processAllResources`, and we stringify that and other `topLevelErrors` and store in the `Err` field of `UpdateErrorMetadata`. 

The `XxxUpdate` structs are used all the way up the stack, and the updates getting there will not have the error field set. So, it feels like it is not the right place. Wdyt?

Reply:
`UpdateErrorMetadata` is also used to keep per-resource metadata for CSDS (there are maps like `ldsMD map[string]UpdateMetadata`. If we add a map there, per-resource metadata will contain a map of errors for all the resources, that also looks weird.

What we really need is tuple. The updates should be `map[string]<Update, error>`. The easier option now is to add the error to the Update struct. Or I can add a separate `map[string]error` for just the errors

Reply:
> Or I can add a separate map[string]error for just the errors

Where were you thinking of adding this? To the `NewXxx()` methods of the `UpdateHandler` interface?

Reply:
To the NewXXX() methods. `NewListeners(updates map[string]ListenerUpdate, errors map[string]error, metadata UpdateMetadata)`

Or, maybe I should add another struct (hmm, actually, one for each type).

```go
struct {
  update ListenerUpdate
  err error
}
```

We already have this in the tests, maybe I can merge them.

Reply:
I also thought about adding the new struct, but might be a whole bunch of new structs :(
Adding it to `NewXxx()` methods might be easier I feel. I'll leave it up to you. Thanks.

Reply:
Done. Added a tuple struct. Please see the new commit.

Reply:
I think the changes look good to me. One small nit though. The name `ListenerUpdateErr` makes it look as though it is a listener update error. Do you think naming it `ListenerUpdateWithErr` or `ListenerUpdateAndErr` or `ListenerUpdateTuple` makes more sense.

Reply:
Renamed to `ListenerUpdateErrTuple` (why not both :smiley_cat:)